### PR TITLE
Fix to support both server and local-state access in same query

### PR DESCRIPTION
**Pull request by akiran**
In this PR, I am trying to fix the errors thrown when both server and local-state data is accessed in same query

Error is happening because of this [line](https://github.com/apollographql/apollo-link-state/blob/master/packages/apollo-link-state/src/index.ts#L76). resolverMap is undefined for query. so error is thrown when resolverMap[fieldName] is accessed. I fixed this by adding {} as fallback to resolverMap

Also rootValue passed to resolver has only server data.
if resolvers are not explicitly defined, apollo-link-state could not return data for client query. 
To fix this, I wrote a utility getDirectivesFromDocument to get a query with client only fields and read data from cache and merged with server data to pass it to resolver.
https://github.com/akiran/apollo-link-state/blob/combine-server-client-query/packages/apollo-link-state/src/index.ts#L85

This POC is working and I was able to get both server and client data with same query.
@peggyrayzis Can you review this and let me know if the direction of this solution is correct.

If the direction is correct, I will handle all the edge cases and write tests.

### CODE REVIEW

**Comment by brunobasto** [comment]
Awesome!!

**Comment by akiran** [comment]
This PR also fixes https://github.com/apollographql/apollo-link-state/issues/187

**Comment by fbartho** [comment]
This looks really cool, I look forward to using this when it's done @akiran. Having some tests would certainly help make sure this functionality isn't forgotten in the future!

**Comment by akiran** [comment]
@fbartho Thanks, I will complete this PR and add tests shortly

**Comment by peggyrayzis** [comment]
Here's the associated issue for `apollo-utilities`: https://github.com/apollographql/apollo-client/issues/2884

**Comment by akiran** [comment]
@peggyrayzis I created PR for getDirectivesFromDocument  https://github.com/apollographql/apollo-client/pull/2974

**Comment by akiran** [comment]
@peggyrayzis I need a merge function in this PR to support nested queries that has both client and server fields. Should I create a new utility in apollo-utilities or can we use lodash.merge ?

**Comment by akiran** [comment]
@peggyrayzis @fbartho Can you review this

File: packages/apollo-link-state/package.json [hunk 1]
```diff
@@ -79,7 +79,8 @@
   },
   "dependencies": {
     "apollo-utilities": "^1.0.6",
-    "graphql-anywhere": "^4.1.0-alpha.0"
+    "graphql-anywhere": "^4.1.0-alpha.0",
+    "lodash": "^4.17.5"
```
Comment (majelbstoat) [inline packages/apollo-link-state/package.json]: If merge is the only thing you're using, you could instead depend on 'lodash.merge'.

Comment (akiran) [reply packages/apollo-link-state/package.json]: @majelbstoat I tried lodash.merge. ES2015 style import didn't work with it.  It is only working only if I use require statement

Comment (peggyrayzis) [reply packages/apollo-link-state/package.json]: Lodash & the TS compiler are weird sometimes 😢 

Try `import merge = require('lodash/merge');` for the submodule and `npm i @types/lodash -D` so we have the proper typings. This is how we do it in `react-apollo`.

**Comment by akiran** [comment]
@peggyrayzis Latest version of apollo-utilities on npm is still 1.0.6.
Can you confirm new version of apollo is released

**Comment by peggyrayzis** [comment]
Oops @akiran, looks like it didn't publish due to the npm outages last week. I'll let you know when it's ready!

**Comment by akiran** [comment]
@peggyrayzis Can you hold off on apollo-utilities release for a while.
After resolving merge conflicts, some recently added tests are failing in this repo.   
This could be an issue with getDirectivesFromDocument. I think, I didn't handled cases where query is merged with mutations.

I will test it and confirm you shortly

**Comment by akiran** [comment]
@peggyrayzis We needn't support this type of query where query and mutation are combined right ?
```const query = gql`
      query SampleQuery {
        user @client
        product
      }
      mutation SampleMutation {
        login @client
        updateUser
      }
    `;
```
When I run a test with this query on getDirectivesFromDocument, I got below error
` Ambiguous GraphQL document: contains 2 operations`

**Comment by akiran** [comment]
@peggyrayzis I think, there is no issue with getDirectivesFromDocument
I wrote additional tests to make sure it works with mutations as well
https://github.com/apollographql/apollo-client/pull/2995
You can release apollo-utilities

I will debug why tests are failing in this repo and fix it

**Comment by peggyrayzis** [comment]
> @peggyrayzis We needn't support this type of query where query and mutation are combined right ?

No, I don't think so. Thanks for writing the tests to confirm it works with mutations!

@evans do you have any idea why the tests would be failing here with the new `resetStore` changes?

File: packages/apollo-link-state/src/index.ts [hunk 2]
```diff
@@ -83,6 +87,8 @@ export const withClientState = (
       };
 
       return new Observable(observer => {
+        const clientQuery = getClientSetsFromDocument(operation.query);
+        const clientData = cache && cache.readQuery({ query: clientQuery });
```
Comment (akiran) [inline packages/apollo-link-state/src/index.ts]: @evans @peggyrayzis readQuery in line 91 is causing the tests to fails.

**Comment by akiran** [comment]
@evans @peggyrayzis I was trying to understand how resetStore and writeDefaults should work for apollo-link-state

When resetStore is triggered, first stored is reseted and then active queries are executed. only after that callback attached to onResetStore will be executed 

But for apollo-link-state, shouldn't we execute onResetStore callback (writeDefaults) before executing active queries. otherwise rootValue will be undefined and all the active query execution will fail

**Comment by peggyrayzis** [comment]
> But for apollo-link-state, shouldn't we execute onResetStore callback (writeDefaults) before executing active queries. otherwise rootValue will be undefined and all the active query execution will fail

Yeah, that was originally what I wanted to do. `client.resetStore()` is tricky because the function both empties the store AND refetches all active queries. I tried splitting up the function [here](https://github.com/apollographql/apollo-client/pull/2843) to allow for the callbacks in between but wasn't able to get the tests to pass. Maybe a second pair of eyes would help?

In the meantime, @evans recently [made a PR](https://github.com/apollographql/apollo-link-state/pull/202) to check for the presence of a resolverMap that didn't require making changes to the client core.

**Comment by akiran** [comment]
@peggyrayzis I was able to fix 2 tests in this your PR 
https://github.com/apollographql/apollo-client/pull/3010

1 more test is still failing

File: packages/apollo-link-state/src/index.ts (hunk 2 above)
Comment (evans) [reply packages/apollo-link-state/src/index.ts]: This client query should be done in the `graphql` call, try changing the query argument [here](https://github.com/apollographql/apollo-link-state/pull/193/files/d48b1c9738c6abfcc5a1add6065b51166ea5025d#diff-cda1c9af8b0b40f97820e2a76de23040R106) to `clientQuery`

Comment (evans) [reply packages/apollo-link-state/src/index.ts]: The reason for the failure is due to the cache being empty on the first read. The client-side data is populated by the `graphql` call. If you could add a comment above that call saying that this is where the client data is fetched, that would be incredible!

**Comment by akiran** [comment]
@peggyrayzis @evans Now the tests are passing
I also upgraded to latest version of apollo-utilities and removed getDirectivesFromDocument logic from this repo

**Comment by akiran** [comment]
@evans It is possible to fallback to cache instead of defaults here ?
https://github.com/apollographql/apollo-link-state/blob/master/packages/apollo-link-state/src/index.ts#L82

I didn't find a easy way to do that.  So I was reading data from  cache with readQuery https://github.com/akiran/apollo-link-state/blob/e02d692cf53cc09ee728877a78ecfb225538cc04/packages/apollo-link-state/src/index.ts#L99
Because of this #L82 will not execute. It either gets data from cache or execute resolver

**Comment by evans** [comment]
@akiran It looks like `getDirectivesFromDocument` doesn't properly respect fragments included in the query. There are probably bugs with the current solution that we are thinking about. Here is the minimal test case:

```
const query = gql`
    fragment client on ClientData {
      bar
    }

    query Mixed {
      foo @client {
        ...client
      }
      bar {
        baz
      }
    }
  `;
```

Returns:

```
      fragment client on ClientData {}

      query Mixed {
        foo @client {
          ...client
        }
      }
```

**Comment by evans** [comment]
I think the best DX is to include all fields of a fragment in the query, no matter their directive status. I think an addition to the `getDirectivesFromDocument` function of apollo-utilities is necessary to indicate if fragments are included.

**Comment by evans** [comment]
@akiran Thank you so much for your investigation, tests, and implementation! I think we've settled on the issue. One of the [tests](https://github.com/apollographql/apollo-link-state/pull/193/commits/47b3dbf0f03d9698c903eb19e5d3c1f3a26aa1d2#diff-7ba323c168c853807c8b48e7eaf2dc7dR200) was incorrectly stubbing server behavior. If you get a chance to take a look at this PR, please do. 

@jbaxleyiii I'm unsure why this [test](https://github.com/apollographql/apollo-link-state/blob/03e45755645ff798b6b068c9300b8a1723fb23c8/packages/apollo-link-state/src/__tests__/advanced.ts#L214-L261) is passing, since apollo-link-state does not pass back the correct data including the client-side data(indicated by a warning) However the watch query is indeed getting the correct values. I think we need to ensure that the correct data is passed back to through the link chain, so we'll need to add a read from cache in our `graphql-anywhere` resolver.

Still on our list before 1.0 is figuring out how to deal with the setting of defaults in the `onResetStore` callback and storing an empty array in the store as a default.

**Comment by wKovacs64** [comment]
Do you know if this PR is blocking a release that includes your #202, @evans?

**Comment by evans** [comment]
@akiran Thank you so much for the tests and research!

@wKovacs64 We're working on permissions now

**Comment by akiran** [comment]
Thanks @evans

**Comment by pvlv** [comment]
@peggyrayzis when update the package in npm?

**Comment by peggyrayzis** [comment]
I have to cut a release of AC first. I should have it out within the next hour or so.
